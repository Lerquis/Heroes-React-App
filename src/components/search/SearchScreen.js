import React, { useMemo } from "react";
import { useNavigate, useLocation } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import { getHeroesByName } from "../../selectors/getHeroesByName";
import { HeroCard } from "../hero/HeroCard";
import queryString from "query-string";

export const SearchScreen = () => {
  // ?Para navegar a otra pagina
  const navigate = useNavigate();
  // ?Importante el location porque este tiene el contenido de la url
  const location = useLocation();

  // ?Importamos una libreria que nos facilita ver la url y saber
  // ?las cosas que busca el usuario, no importa que tan grande sea la
  // ?url, siempre la divide en partes de busqueda
  // ?Hacemos destructuring, ya que la url nos da nuestro dato de
  // ?busqueda en un objeto donde se almacena 'q'
  const { q = "" } = queryString.parse(location.search);

  // ?Agregamos nuestro hook para manejar el formulario
  const [{ searchText }, handleInputChange, reset] = useForm({
    // ?De default le damos el valor que tiene q
    searchText: q,
  });

  // ?Obtenemos los resultados mediante esta funcion, donde de
  // ?parametro le enviamos la busqueda del usuario
  // ?lo hacemos en un useMemo, ya que este cada que se escribe
  // ?renderiza el componente, y por ende no es tan eficiente
  // ?en un use memo, manda la funcion cuando cambie 'q'
  const listaResultados = useMemo(() => getHeroesByName(q), [q]);

  // ?Cuando se le de al boton de submit
  const handleSearch = (e) => {
    // ?prevenimos que se recargue la pagina
    e.preventDefault();

    // ?Lo enviamos a la pagina con los resultados
    navigate(`?q=${searchText}`);
  };

  return (
    <>
      <h1>Search for a hero!</h1>
      <hr></hr>

      <div className="row">
        <div className="col-5">
          <form
            onSubmit={(e) => {
              handleSearch(e);
            }}
          >
            <input
              type="text"
              onChange={handleInputChange}
              placeholder="Search for a hero"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
            />

            <button type="submit" className="btn btn-outline-primary mt-3">
              Search
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Resultados</h4>
          <hr></hr>

          {/*
          La sintaxis de abajo quiere decir que si no hay busqueda, muestre 
          un div que diga que busca un hero. Sino, si el length de
          los resultados es 0 y hay busqueda muestre un div diciendo
          que no hay resultados*/}
          {q === "" ? (
            <div className="alert alert-info">Search a hero!!</div>
          ) : (
            listaResultados.length === 0 && (
              <div className="alert alert-danger">
                No hay resultados para: {q}
              </div>
            )
          )}

          {listaResultados.map((hero) => {
            return <HeroCard key={hero.id} {...hero} />;
          })}
        </div>
      </div>
    </>
  );
};
