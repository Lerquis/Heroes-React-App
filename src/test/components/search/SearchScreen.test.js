import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

// !IMPORTANTE QUE EL NOMBRE SEA mockNavigate, que tenga el MOCK. Sino
// !va a tener error
const mockNavigate = jest.fn();
// ?tenemos que simular el useNavigate. Eso forma parte de react-router
// ?entonces simulamos mediante un mock, una funcion que nos devuelve
// ?todas las funciones de jest mediante un spread (...) y podemos
// ?sobreescribir la funcion que ocupamos, que es el useNavigate
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

describe("Pruebas en el <SearchScreen/>", () => {
  test("Debe de mostrarse correctamente con valores por defecto", () => {
    // ?Tiene que estar adentro del memoryRouter lo que deseamos
    // ?probar
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    // ?Esperamos que el texto que tenga esa clase, sea el escrito
    expect(wrapper.find(".alert-info").text().trim()).toBe("Search a hero!!");
  });

  test("Debe de mostrar a 'Batman' y el input con el valor del queryString", () => {
    // ?Tiene que estar adentro del memoryRouter lo que deseamos
    // ?probar, en este caso le agregamos el queryString para buscar
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <SearchScreen />
      </MemoryRouter>
    );
    // ?Esperamos que el valor que tiene el input sea de batman
    expect(wrapper.find("input").prop("value")).toBe("batman");
  });

  test("Debe de mostrar el mensaje de error si no encuentra el hero", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman123"]}>
        <SearchScreen />
      </MemoryRouter>
    );
    // ?Ahora lo hacemos tirando un error de que no encontro un hero
    expect(wrapper.find(".alert-danger").text().trim()).toBe(
      "No hay resultados para: batman123"
    );
  });

  test("Debe de llamar el navigate a la nueva pantalla", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <SearchScreen />
      </MemoryRouter>
    );
    // ?Buscamos el input y simulamos un cambio en el target que
    // ?tenga ese name y le asignamos ese value
    wrapper
      .find("input")
      .simulate("change", { target: { name: "searchText", value: "batman" } });

    // ?Ahora buscamos el form y seleccionamos su propiedad de la
    // ?funcion de submit
    // ?Como el onSubmit tiene de parametro un evento y tenemos el prevent
    // ?de ese evento, tenemos que enviarselo para que no tire error
    wrapper.find("form").prop("onSubmit")({ preventDefault: () => {} });

    // ?Esperamos que el mockNavigate, es decir, el useNavigate, se
    // ?haya llamado
    expect(mockNavigate).toHaveBeenCalled();

    // ?Esperamos que el mockNavigate se haya llamado con ?q=batman
    // ?que es el valor que le pusimos al input
    expect(mockNavigate).toHaveBeenCalledWith("?q=batman");
  });
});
