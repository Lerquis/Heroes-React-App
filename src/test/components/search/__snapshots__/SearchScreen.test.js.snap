// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <SearchScreen/> Debe de mostrarse correctamente con valores por defecto 1`] = `
Array [
  <h1>
    Search for a hero!
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search for a hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-3"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero!!
      </div>
    </div>
  </div>,
]
`;
